### PR TITLE
Add shared SQLite model admission

### PR DESCRIPTION
**Pull request by jackbravo**
## Summary
- coordinate global model slots and queue admission through SQLite WAL
- support multiple web workers without multiplying model concurrency
- add leases and restart recovery for interrupted executions
- expose queue position, wait estimates, and active slot status in the UI
- make hybrid embedding-server lifecycle safe for multiple web workers

## Validation
- `UV_CACHE_DIR=/tmp/codex-uv-cache uv run python -m unittest discover -s tests -q` (141 passed)
- `UV_CACHE_DIR=/tmp/codex-uv-cache uv run ruff check human_eval tests`
- `git diff --check`

The implementation is based on the existing admission-queue visibility work and is isolated from the unrelated local checkout.

### CODE REVIEW

**Comment by jackbravo** [comment]
## Review

**Verified locally**: 141 tests pass (8.2s), `ruff check` clean, `git diff --check` clean.

The design is sound: atomic claim + `started` insert in one `BEGIN IMMEDIATE` transaction, queue-position tie-break matching claim order, DB polling as fallback for the in-process wakeup queue, and startup recovery for both expired leases and pre-v4 `started` runs are all done right. Queue position/wait estimates in the UI, the SSE `queue` event, `Retry-After` on 503, and the `--workers` guard for `DOF_MANAGE_EMBED_SERVER=false` round it out nicely.

### Issues found (1 and 2 addressed in 7bd29b2)

**1. Global correctness silently depends on uniform env config** (medium, now documented). `claim_next_run`, `model_activity`, and `initialize_model_slots` all filter by `slot_id <= <this process's concurrency>`, and each process's `create_run` enforces its own `queue_capacity`. If two web processes run with different `DOF_MODEL_CONCURRENCY`/`DOF_QUEUE_CAPACITY`, the effective global limit is the max of them and nothing warns. → Fixed by documenting the requirement in `README.md`. A follow-up could store concurrency in `schema_meta` on first init and refuse on mismatch.

**2. Lease expiry could fail a still-running execution and kill the worker thread** (now fixed). If a heartbeat stalled (or after `close()`'s `expire_model_leases` while `execute()` was still in flight past the join timeout), a starting process could mark the run `failed` and free the slot while the old executor kept running. The old worker's late `succeeded` then hit the transition guard in `append_event` and the `ValueError` propagated, killing the scheduler worker thread. → `_append_event_if_open` now catches the rejected transition, logs, and drops the late event so the recovery-written terminal state wins and the worker keeps serving. Regression test `test_worker_survives_when_recovery_wins_the_lease_race` covers it (142 tests pass). Note the freed slot can still briefly exceed intended model concurrency while the orphaned execution finishes — unavoidable without a cancellation mechanism; acceptable.

**3. `create_run`'s active-run check is hidden behind `queue_capacity is not None`** (minor API smell). Passing `queue_capacity=None` silently disables the per-user active-run check too. Consider making them independent parameters.

**4. Daily-quota check remains outside the admission transaction** (preexisting). Two processes can both pass the daily-limit check concurrently; low impact at 1 question/day, but the old comment acknowledging the single-process assumption was removed, so worth tracking.

**5. Multi-thread executor use is new and only implicitly safe.** With `scheduler_workers > 1` (defaults to `model_concurrency`), `AgentRunExecutor.execute` runs on several threads. Embedder creation and `embed_query` are locked and `_backend()`/`DofRetriever` are per-call, so it looks OK — but the parallelism lands on one llama-server, which needs `-np N` to match (README already covers this).

**Nits**: workers poll every 0.25s even when idle (fine; could back off); `estimated_wait_seconds` ignores remaining time of active runs (fine as a rough estimate).

**Verdict: approve** — 3–5 are follow-ups, not blockers.

**Comment by jackbravo** [comment]
Follow-up review items addressed in `10fb0a7`:

- `create_run` now exposes `enforce_active_run` independently from `queue_capacity`; the service enables it explicitly.
- Rolling daily quota is rechecked inside the same `BEGIN IMMEDIATE` admission transaction, with a dedicated `DailyQuotaConflict`.
- Added regression coverage for independent admission constraints and a single service using two scheduler workers/model slots concurrently.
- Updated the deployment notes to identify this Mac as development/benchmark-only; production runs on the moved server machine.

The existing lease-race fix and shared scheduler configuration documentation remain in `7bd29b2`.

**Comment by jackbravo** [comment]
Readiness fixes and latest `main` integration are in `3c8a363`:

- resolved conflicts in `docs/production-roadmap.md`, `human_eval/service.py`, and `tests/test_human_eval.py`; retained main’s privacy-safe admission logs and compact-tool/premise-status work
- multi-worker Uvicorn now honors `DOF_WEB_HOST` and `DOF_WEB_PORT`
- claim polling now reaps leases that expire after replacement workers have already started
- shutdown lets active leases expire naturally instead of freeing a slot while the old inference may still be running
- late-result handling catches only `InvalidRunTransition`, rather than swallowing every `ValueError`
- added regression tests for launch configuration, post-startup lease expiry, and non-transition errors

Validation after merging main: 155 tests pass, Ruff is clean, and diff checks pass. GitHub now reports the PR mergeable; it remains blocked only on the required review. GitHub Actions should be introduced in a separate PR because this repository currently has no workflow and CI requires its own runtime/cache/permissions decisions.

File: human_eval/service.py [hunk 1]
```diff
@@ -144,74 +173,95 @@ def submit(
...
+                    provenance=self.provenance_factory(),
+                    enforce_active_run=True,
+                    queue_capacity=self.queue.maxsize,
+                    daily_question_limit=(
+                        daily_question_limit
+                        if not admin and daily_question_limit >= 1
+                        else None
+                    ),
+                    daily_since=daily_since,
+                )
```
[the remark this answers is not in the document]
Comment (jackbravo) [reply human_eval/service.py:248]: Fixed in `1421a7d`. `EvaluationStore.create_run` now compares question, cutoff, and hop count inside the same `BEGIN IMMEDIATE` transaction before returning an existing idempotent run. A dedicated store conflict is mapped back to `IdempotencyConflictError`, including when both processes missed the preliminary lookup. Added store- and service-level regressions for this loser path.

File: human_eval/service.py [hunk 2]
```diff
@@ -278,53 +361,105 @@ def delete_run(self, run_id: str) -> None:
...
-                    self.queue.task_done()
+                    if notification != "__poll__":
+                        self.queue.task_done()
         finally:
-            if self._closing.is_set():
+            if self._closing.is_set() and not any(
+                worker.is_alive() and worker is not threading.current_thread()
+                for worker in self.workers
+            ):
                 self._close_executor()
```
[the remark this answers is not in the document]
Comment (jackbravo) [reply human_eval/service.py]: Fixed in `1421a7d`. Worker shutdown now uses a lock-protected active-worker counter; the unique transition to zero invokes the already-idempotent executor close hook. This removes the mutual `is_alive()` race. Added a two-worker timed-out-shutdown regression that verifies cleanup waits for both in-flight calls and then closes the executor.

File: human_eval/store.py [hunk 3]
```diff
@@ -210,6 +271,215 @@ def create_run(
...
+        concurrency: int,
+        lease_seconds: float,
+    ) -> tuple[str, int] | None:
+        """Atomically claim the oldest queued run and one model slot."""
+        if concurrency < 1 or lease_seconds <= 0:
+            raise ValueError("invalid scheduler limits")
+        now = utc_now()
+        lease_until = self._lease_until(lease_seconds)
+        with self._connect() as connection:
+            connection.execute("BEGIN IMMEDIATE")
```
[the remark this answers is not in the document]
Comment (jackbravo) [reply human_eval/store.py]: Fixed in `1421a7d`. `claim_next_run` now acquires `BEGIN IMMEDIATE` before calculating both `now` and `lease_until`. The regression holds a competing SQLite write lock longer than the requested lease and verifies the committed deadline still begins after lock acquisition.

File: human_eval/store.py [hunk 4]
```diff
@@ -210,6 +271,215 @@ def create_run(
...
+        lease_seconds: float,
+    ) -> bool:
+        if lease_seconds <= 0:
+            raise ValueError("lease_seconds must be positive")
+        with self._connect() as connection:
+            cursor = connection.execute(
+                "UPDATE model_slots SET lease_until = ? WHERE slot_id = ? "
+                "AND run_id = ? AND worker_id = ?",
+                (self._lease_until(lease_seconds), slot_id, run_id, worker_id),
+            )
```
[the remark this answers is not in the document]
Comment (jackbravo) [reply human_eval/store.py]: Fixed in `1421a7d`. `renew_model_slot` now explicitly acquires `BEGIN IMMEDIATE` and calculates the renewed deadline only after obtaining that lock. A matching lock-contention regression verifies the renewed lease is fresh after the blocked writer releases.

File: tests/test_human_eval.py [hunk 5]
```diff
@@ -965,6 +1085,111 @@ def test_queue_full_rejection_is_logged(self):
...
+        first_service.start()
+        second_service.start()
+        try:
+            first = first_service.submit(
+                RunRequest("primera", client_request_id="shared-1"),
+                user_id="shared-u1",
+                admin=True,
+            )
+            self.assertTrue(first_executor.started.wait(timeout=1))
+            second = second_service.submit(
```
[the remark this answers is not in the document]
Comment (jackbravo) [reply tests/test_human_eval.py:1384]: Fixed in `1421a7d`. The test now starts `second_service` only after `first_executor.started`, making ownership of the first slot deterministic while preserving the cross-service concurrency assertion.

**Comment by jackbravo** [comment]
Also addressed the two suppressed Copilot findings in `1421a7d`: progress writes now atomically require the run to still be `started` and the caller to own an unexpired model-slot lease, and heartbeat cadence is always one-third of the configured lease (including sub-100ms test leases). The recovered-run regression now confirms late progress is dropped. Full suite: 162 tests pass; Ruff and diff checks pass.

File: human_eval/service.py [hunk 6]
```diff
@@ -278,53 +368,107 @@ def delete_run(self, run_id: str) -> None:
...
+                            self._append_event_if_open(
+                                run_id,
+                                "failed",
+                                {
+                                    "code": "internal_error",
+                                    "message": "La ejecución no pudo completarse.",
+                                },
+                            )
+                        else:
+                            self._append_event_if_open(run_id, "succeeded", result)
```
[the remark this answers is not in the document]
Comment (jackbravo) [reply human_eval/service.py]: Fixed in `d02dd52`. Terminal persistence now uses `append_terminal_event_if_claimed`, which acquires `BEGIN IMMEDIATE` and atomically verifies the latest state is `started` plus the exact slot/run/worker lease is still unexpired before inserting `succeeded` or `failed`. A rejected stale terminal result leaves the expired slot for normal recovery instead of releasing it into an orphaned started state. `test_expired_unreaped_lease_cannot_commit_success` covers the sole-scheduler case. I also prevented heartbeats from resurrecting an already-expired lease.

File: human_eval/store.py [hunk 7]
```diff
@@ -210,6 +284,221 @@ def create_run(
...
+        now = utc_now()
+        recovered = 0
+        with self._connect() as connection:
+            connection.execute("BEGIN IMMEDIATE")
+            runs = connection.execute(
+                "SELECT r.run_id, e.sequence FROM runs r JOIN run_events e "
+                "ON e.run_id = r.run_id WHERE e.sequence = (SELECT MAX(e2.sequence) "
+                "FROM run_events e2 WHERE e2.run_id = r.run_id) "
+                "AND e.event_type = 'started' AND NOT EXISTS ("
+                "SELECT 1 FROM model_slots s WHERE s.run_id = r.run_id)"
```
[the remark this answers is not in the document]
Comment (jackbravo) [reply human_eval/store.py]: Fixed in `d02dd52`. Unclaimed-started recovery is no longer run on every service startup. It now runs only while atomically upgrading a database whose schema version is older than v4; the version is re-read under `BEGIN IMMEDIATE` so concurrent initializers cannot apply stale migration logic to a newly started v4 seed. Tests cover both recovery of a pre-v4 orphan and preservation/completion of a live current-schema `seed_live_run` while a web service starts.

**Comment by jackbravo** [comment]
Second Copilot round addressed in `d02dd52`. In addition to the two requested fixes, lease renewal now refuses to resurrect an already-expired lease. Validation: 166 tests pass, the concurrency-focused human-evaluation suite passed repeatedly, Ruff passes, and `git diff --check` passes.

File: human_eval/store.py [hunk 8]
```diff
@@ -122,17 +151,32 @@ def initialize(self) -> None:
...
                 raise RuntimeError(
                     f"unsupported evaluation schema {current[0]!r}; expected {SCHEMA_VERSION}"
                 )
             connection.executescript(SCHEMA)
             self._migrate_columns(connection)
+            # executescript may commit while creating missing tables. Re-enter
+            # a write transaction and re-read the version so only one process
+            # performs v4 orphan recovery during a concurrent deployment.
+            connection.commit()
+            connection.execute("BEGIN IMMEDIATE")
```
[the remark this answers is not in the document]
Comment (jackbravo) [reply human_eval/store.py]: Fixed in `9dec237`. Initialization now commits only the bootstrap `schema_meta` creation, acquires `BEGIN IMMEDIATE`, and keeps version inspection, schema creation, `_migrate_columns`, legacy orphan recovery, and the version update inside that single write transaction. Because `sqlite3.executescript` would implicitly commit, schema statements are executed individually using `sqlite3.complete_statement`. The new regression verifies `_migrate_columns` runs while a competing `BEGIN IMMEDIATE` is locked out; existing v1/v2/v3 migration tests also pass.

**Comment by jackbravo** [comment]
The three suppressed findings from the latest Copilot review are also fixed in `9dec237`:

- the stream removes stale queue status on the first progress event and on a new explicit `running` SSE event;
- queue SSE messages are emitted only from a complete snapshot, so a queued/running read race cannot produce position 0 or 0-of-0 slots;
- `Retry-After` now estimates the next completion (one recent average duration), not the time to drain the full queue.

Added focused regressions for migration locking, complete queue snapshots, running-state UI handling, and next-completion retry timing. Full validation: 170 tests pass; Ruff and diff checks pass.

**Comment by jackbravo** [comment]
Fixed all five full-review findings in 1cdf21186121dc2337b3c5468b07efeea7013409.

The three suppressed Copilot comments are addressed:
- Shutdown joins share one monotonic deadline instead of multiplying the timeout by scheduler worker count.
- Wait estimates subtract available slots; HTML and SSE consistently show “menos de 1 min” for zero/near-zero waits.
- Completed runs release their owned slot even when shutdown begins between terminal persistence and cleanup.

The two additional reproduced races are also fixed:
- A late terminal result for a recovered-and-deleted run is rejected without killing its scheduler. A regression verifies the next question completes.
- Live seed creation and its started event commit atomically, so web schedulers cannot claim the seed between those operations. Maintenance execution remains independently managed.

Validation: 176 tests passed, including six new regressions (`uv run --no-sync python -m unittest discover -s tests -q`); Ruff passed for human_eval, tests, and the modified seed script; git diff --check passed.

All existing inline review threads are already resolved. The three suppressed comments are embedded in Copilot’s review summary and have no separate resolvable thread; this comment records their fixes.

File: human_eval/service.py [hunk 9]
```diff
@@ -278,53 +367,112 @@ def delete_run(self, run_id: str) -> None:
...
+                except queue.Empty:
+                    notification = "__poll__"
+                try:
+                    if notification is None or self._closing.is_set():
                         return
+                    claim = self.store.claim_next_run(
+                        worker_id=self.worker_id,
+                        concurrency=self.model_concurrency,
+                        lease_seconds=self.lease_seconds,
+                    )
```
[the remark this answers is not in the document]
Comment (jackbravo) [reply human_eval/service.py]: Fixed in 55cb39e. The closing check and claim now hold the same _write_lock used by close(), so claims cannot commit after the closing transition. Scheduler claims also use a short SQLite busy timeout bounded by the poll interval, lease interval, and shutdown timeout, releasing the local lock between retries. Added regression coverage for shared-lock ownership and suppression of claims after shutdown. Full suite: 182 tests pass; Ruff and diff checks pass.

File: human_eval/service.py [hunk 10]
```diff
@@ -278,53 +367,112 @@ def delete_run(self, run_id: str) -> None:
...
+        interval = self.lease_seconds / 3
+        while not stop.wait(interval):
+            if self._closing.is_set():
+                return
+            if not self.store.renew_model_slot(
+                run_id=run_id,
+                slot_id=slot_id,
+                worker_id=self.worker_id,
+                lease_seconds=self.lease_seconds,
+            ):
```
[the remark this answers is not in the document]
Comment (jackbravo) [reply human_eval/service.py]: Fixed in 55cb39e. Heartbeats retry SQLITE_BUSY/SQLITE_LOCKED (including extended codes) with bounded exponential backoff within a monotonic lease budget. Each renewal has a short busy timeout capped by the remaining budget; successful renewal resets the budget, and the store still rejects expired leases. Non-transient database errors are not hidden. Regressions cover recovery and successful completion, deadline exhaustion, non-transient errors, and a real SQLite lock with a zero busy timeout. Full suite: 182 tests pass; Ruff and diff checks pass.

File: human_eval/service.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (jackbravo) [reply human_eval/service.py]: Fixed in 55cb39e. Each scheduler iteration catches retryable SQLITE_BUSY/SQLITE_LOCKED errors, logs the failure, and waits interruptibly before continuing. Queue notification accounting remains in finally, and persisted claims remain available to normal lease recovery if a later store operation fails. Non-transient errors still propagate. A regression injects a failed claim and verifies that the same scheduler completes the queued run afterward. Full suite: 182 tests pass; Ruff and diff checks pass.

File: human_eval/service.py [hunk 11]
```diff
@@ -278,53 +376,156 @@ def delete_run(self, run_id: str) -> None:
...
-                            payload: self._append_progress_if_open(
-                                run_id, event_type, payload
-                            ),
-                        )
-                    except PublicExecutionError as exc:
-                        if exc.__cause__ is not None:
-                            LOGGER.exception(
-                                "human-evaluation run %s failed with %s",
+                        try:
+                            result = self.executor.execute(
```
[the remark this answers is not in the document]
Comment (jackbravo) [reply human_eval/service.py:422]: Fixed in `2952831`. The heartbeat now renews the lease immediately after each claim (first `stop.wait(0)` iteration) instead of waiting a full interval, so a stalled claim→execution gap can no longer outlive the lease before the first renewal. A regression with `lease_seconds=30` asserts the first renewal lands within 3 s of the claim rather than after the 10 s interval. If that first renewal fails, the fenced terminal write still rejects the result, so an orphaned execution can never persist over a recovered run.

**Comment by jackbravo** [comment]
The latest round of Copilot findings is addressed in `2952831`:

- **Terminal write abandoned on transient contention** (`service.py:553`): terminal persistence now goes through `_append_terminal_with_retry`, which retries retryable `SQLITE_BUSY`/`SQLITE_LOCKED` with bounded backoff inside one heartbeat interval while the heartbeat is still renewing the lease. The fenced store write still rejects the result if the lease was genuinely lost, so recovery semantics are unchanged.
- **Progress contention failed healthy runs** (`service.py:584`): `_append_progress_if_open` now drops only the affected progress event (with a warning) on retryable store errors instead of letting the exception propagate through `on_progress` and abort the execution as `internal_error`. Non-transient database errors still propagate and fail the run.

Regressions added: transient busy error on the terminal write is retried and the run succeeds with its slot released; a transient busy error on progress is dropped while the run succeeds; a non-transient progress error still fails the run; the first lease renewal after a claim is immediate (see the inline reply for the claim→execution gap fix).

Validation: 186 tests pass (five consecutive runs of the concurrency-focused suite are stable), Ruff passes for `human_eval` and `tests`, and `git diff --check` passes.

**Comment by jackbravo** [comment]
Closing the design exploration here: the review rounds on this PR repeatedly surfaced edge cases in the lease/heartbeat/fencing protocol, which exists only because execution lifetime was coupled to web process lifetime. Rather than continue patching that protocol, we redesigned around it.

Superseded by #85, which keeps this PR's durable work (transactional admission, persistent queue, queue-position/wait UI, SSE queue events, `Retry-After`, multi-worker uvicorn) and replaces the lease machinery with a singleton scheduler process: one flock, startup recovery of orphaned `started` runs, an in-process concurrency pool, and systemd-supervised restart. Schema stays at v3.

Thank you to the reviewers (human and Copilot) — every finding here informed the invariants the replacement enforces structurally.
